abi.md

```diff
@@ -0,0 +1,147 @@
+similar to an application binary interface (ABI) which enables predictability of
+the output of key constructs necessary for the interoperability between circuits
+described in FIRRTL and between other languages and FIRRTL output.
+
+This document describes multiple versions of the ABI, specifically calling
+specific changes in later versions.  It is expected that a conforming FIRRTL
+compiler can lower to all specified ABIs.  This mechanism exists to allow
+improved representations when using tools which have better Verilog support and
+allow migration of existing development flows to significant representational
+changes.
+
+# FIRRTL System Verilog Interface
+
+To use a circuit describe in FIRRTL in a predictable way, the mapping of certain
+behaviors and boundary constructs in FIRRTL to System Verilog must be defined.
+Where possible this ABI does not impose constraints on implementation,
+concerning itself primarily with the boundaries of the circuit.
+
+## On Modules
+
+### The Circuit and Top Level
+
+The top level module, as specified by the circuit name, shall be present as a
+System Verilog module of the same name.  This firrtl module is considered a
+"public" module and subject to the lowering constraints for public modules.
+
+### External Modules
+
+An external module may be presumed to exist following the lowering constraints
+for public modules.  The module shall exist with a verilog name matching the 
+defname value or, lacking that, the module name.
+
+###  Public Modules
+
+Any module considered a "public" module shall be implemented in Verilog in a
+consistent way.  Any public module shall exist as a Verilog module of the same
+name.
+
+Each public module with definitions (e.g. not external modules) shall be placed
+in a file with the same name.  No assumption is made of the filename of an
+implementation of an external module, it is the user's job to include files in
+such a way in their tools to resolve the name.
+
+### Port Lowering ABIv1
+
+Ports are generally lowered to netlist types, except where Verilog's type system
+prevents it.
+
+Ports of integer types shall be lowered to netlist ports (`wire`{.verilog}) as a
+packed vector of equivalent size.  For example, consider the following FIRRTL:
+
+```firrtl
+circuit Top :
+  module Top :
+    output out: UInt<16>
+    input b: UInt<32>
+```
+
+This is lowered to the following Verilog:
+
+```verilog
+module Top(
+    output wire [15:0] out,
+    input wire [31:0] in
+);
+```
+
+Ports of aggregate type shall be lowered according to the "Aggregate Type
+Lowering" description in the firrtl spec.
+
+Ports of enum type shall be lowered to ports of a type consistent with the rules
+for lowering firrtl enum types to Verilog.
+
+Ports of ref type shall be lowered to a Verilog macro of the form `` `define
+ref_<circuit name>_<module name>_<portname> <internal path from module>`` in a
+file with name `ref_<circuit name>_<module name>.sv`.
+
+### Port Lowering ABIv2
```

Comment:
It looks like we're defining multiple ABIs in this document. If that is the case, we should make sure to motivate why we need more than one. (Especially for the initial version of the doc).

Reply:
there are different versions used even in the same spec, for public vs internal modules

Reply:
internal modules have no ABI constraints.
v1 is essentially documenting what the contract is now.
v2 is aggregate-preserving.
They are not remotely compatible.  These things can be stated explicitly.